File: CS101/Co-Pilot/PopularCategory.py
'''
This problem is given two string parameters called words, and categories where the kth word in words corresponds to the kth category in categories, meaning the word is in that category. This function calculates the most popular category, returning a string of sorted words in that category. If there is a tie, then of the categories that are tied, return the words from the category first in alphabetical order.
It is possible that there are not the same number of words in the string words or categories. If that is the case, only use the words that have a corresponding word in the other string.

Here is an example with one category with the most words, and also the number of words in the two strings is not equal.

words =  "ten purple eight owl red blue five fish yellow cat dog six nine bird blue"
categories = "number color number animal color color number animal color animal animal number number" 
Here words has 15 words and categories has 13 words. We ignore the last two words in words. With the remaining words we see there are 3 categories: number has five words, animal has four words and color has four words. Since the number category has the most words we return those words in sorted order, returning: "eight five nine six ten"

Here is an example with categories in a tie.

words = "red blue five fish yellow cat dog six nine bird blue"
categories = "color color number animal color animal animal number number" 
  
You can see that red is in the color category, blue is in the color category, five is in the number category and fish is in the animal category. Note fish is in index position 3 in words and animal is its category since animal is in index position 3 in categories.

Here we see there are three words in the color category, three words in the number category and three words in the animal category. The most number in any category is three and all three categories have three items in their category so it is a three-way tie. Since animal comes first in alphabetical order, before number and color, the return value is all the words in the animal category, in sorted order, or "cat dog fish".

Write function popular that has two string parameters named words and categories, where the kth word in words corresponds to the kth word in categories. This function returns a string of the sorted words from the largest category. If there is a tie, then from the categories that tied, choose the category that comes first in alphabetical order.

words is of length 1 or more characters, and is only lowercase alphabetic characters or blanks. It starts and ends with an alphabetic character, and if there are blanks, there cannot be two adjacent blanks.
categories is of length 1 or more characters, and is only lowercase alphabetic characters or blanks. It starts and ends with an alphabetic character, and if there are blanks, there cannot be two adjacent blanks.
There can be duplicate words in words or categories. Duplicate words should be counted.
It is possible that a word could be categorized in more than one category.
'''

def popular(words, categories):
    # split the words and categories into lists
    words = words.split()
    categories = categories.split()
    # create a dictionary to store the categories and their words
    cat_dict = {}
    # loop through the categories and words
    for i in range(len(categories)):
        # if the category is not in the dictionary, add it
        if categories[i] not in cat_dict:
            cat_dict[categories[i]] = [words[i]]
        # if the category is in the dictionary, append the word to the list
        else:
            cat_dict[categories[i]].append(words[i])
    # create a list to store the number of words in each category
    cat_count = []
    # loop through the dictionary
    for key in cat_dict:
        # append the length of the list of words in each category to the list
        cat_count.append(len(cat_dict[key]))
    # find the max number of words in a category
    max_count = max(cat_count)
    # create a list to store the categories with the max number of words
    max_cat = []
    # loop through the dictionary
    for key in cat_dict:
        # if the length of the list of words in each category is the max number of words, append the category to the list
        if len(cat_dict[key]) == max_count:
            max_cat.append(key)
    # if there is only one category with the max number of words, return the words in that category in sorted order
    if len(max_cat) == 1:
        return ' '.join(sorted(cat_dict[max_cat[0]]))
    # if there is a tie, return the words in the category that comes first in alphabetical order
    else:
        return ' '.join(sorted(cat_dict[max_cat[0]]))

File: CS101/Co-Pilot/PopularCategory.java
/*
Given a phrase, your task is to calculate which word length occurs the most times. Return a string that has three components.
the number of words for the most times for a length
a colon
the length of words that occur the most. If there is a tie with more than one length that occurs the most, the tied lengths are in numeric sorted order separated by dashes.
Here is an example with one most value. Consider the phrase "this is how we count length in a line"

There are:

1 word of length 1: a
3 words of length 2: is we in
1 word of length 3: how
2 words of length 4: this line
1 word of length 5: count
1 word of length 6: length
The length with the most words is length 2, there are 3 words of length 2. Return the string "3:2"
Here is an example with a tie for the length with the most words. Consider the phrase "one two three four five six seven eight nine ten eleven twelve thirteen twenty thirty forty"

There are:

4 words of length 3: one two six ten
3 words of length 4: four five nine
4 words of length 5: three seven eight forty
4 words of length 6: eleven twelve twenty thirty
1 word of length 8: thirteen
The word length that occurs the most is a tie with length 3, length 5 and length 6. They each have 4 words, which is the most.
The return value would be the most number of words, 4, followed by ":", and followed by the sorted lengths that have that many words, 3, 5 and 6, with a colon as separator. The return value is the string "4:3-5-6".

Write function count that has one string parameter named phrase. This function returns a string in the format described above that indicates the length that has the most words (or all the ones that tied, in sorted numeric order and separated by dashes) and how many words was the most. Use the format above.
*/

public class PopularCategory {
    public String count(String phrase)
    {
        String[] words = phrase.split(" ");
        int[] counts = new int[words.length];
        for (int i = 0; i < words.length; i++)
        {
            counts[i] = words[i].length();
        }
        int[] counts2 = new int[counts.length];
        for (int i = 0; i < counts.length; i++)
        {
            counts2[i] = counts[i];
        }
        int max = 0;
        for (int i = 0; i < counts.length; i++)
        {
            if (counts[i] > max)
            {
                max = counts[i];
            }
        }
        int[] maxes = new int[counts.length];
        int count = 0;
        for (int i = 0; i < counts.length; i++)
        {
            if (counts[i] == max)
            {
                maxes[count] = counts[i];
                count++;
            }
        }
        int[] maxes2 = new int[count];
        for (int i = 0; i < count; i++)
        {
            maxes2[i] = maxes[i];
        }
        for (int i = 0; i < maxes2.length; i++)
        {
            for (int j = i + 1; j < maxes2.length; j++)
            {
                if (maxes2[i] > maxes2[j])
                {
                    int temp = maxes2[i];
                    maxes2[i] = maxes2[j];
                    maxes2[j] = temp;
                }
            }
        }
        String result = count + ":";
        for (int i = 0; i < maxes2.length; i++)
        {
            result += maxes2[i];
            if (i < maxes2.length - 1)
            {
                result += "-";
            }
        }
        return result;
    }
}